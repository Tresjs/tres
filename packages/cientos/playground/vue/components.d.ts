/* eslint-disable */
// @ts-nocheck
// Generated by unplugin-vue-components
// Read more: https://github.com/vuejs/core/pull/3399
// biome-ignore lint: disable
export {}

/* prettier-ignore */
declare module 'vue' {
  export interface GlobalComponents {
    AkuAku: typeof import('./src/components/gltf/BlenderCube.vue')['default']
    BlenderCube: typeof import('./src/components/gltf/BlenderCube.vue')['default']
    FboCube: typeof import('./src/components/FboCube.vue')['default']
    Gltf: typeof import('./src/components/gltf/index.vue')['default']
    GraphPane: typeof import('./src/components/GraphPane.vue')['default']
    Jeep: typeof import('./src/components/fbx/Jeep.vue')['default']
    ModelsDemo: typeof import('./src/components/ModelsDemo.vue')['default']
    OverlayInfo: typeof import('./src/components/OverlayInfo.vue')['default']
    RouterLink: typeof import('vue-router')['RouterLink']
    RouterView: typeof import('vue-router')['RouterView']
    StarsForTest: typeof import('./src/components/StarsForTest.vue')['default']
  }
}
