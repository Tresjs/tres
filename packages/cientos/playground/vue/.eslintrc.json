{
  "extends": "@tresjs/eslint-config-vue",
  "rules": {
    "no-console": "off",
    "vue/attribute-hyphenation": [2, "always", {
      "ignore": [
        "shadow-mapSize-width",
        "shadow-mapSize-height",
        "distortionMap",
        "material-uniforms-mieCoefficient-value",
        "material-uniforms-mieDirectionalG-value",
        "material-uniforms-sunPosition-value"
      ]
    }]
  }
}
