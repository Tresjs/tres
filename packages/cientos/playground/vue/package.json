{
  "name": "playground",
  "type": "module",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "dev": "vite --host",
    "build": "vue-tsc && vite build",
    "preview": "vite preview"
  },
  "dependencies": {
    "@tresjs/core": "5.0.0",
    "vue-router": "^4.5.1"
  },
  "devDependencies": {
    "@tresjs/leches": "^0.14.1",
    "unplugin-auto-import": "^20.1.0",
    "unplugin-vue-components": "^29.0.0",
    "vite-plugin-glsl": "^1.5.1",
    "vite-plugin-qrcode": "^0.3.0",
    "vue-tsc": "^3.0.7"
  }
}
