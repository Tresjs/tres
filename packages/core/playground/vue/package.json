{
  "name": "playground",
  "type": "module",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "dev": "vite --host",
    "build": "vite build",
    "preview": "vite preview" 
  },
  "dependencies": {
    "@tresjs/cientos": "5.0.0",
    "@tresjs/core": "workspace:^",
    "@tresjs/leches": "https://pkg.pr.new/@tresjs/leches@b34e795",
    "vue-router": "^4.5.1"
  },
  "devDependencies": {
    "@tweakpane/plugin-essentials": "^0.2.1",
    "unplugin-auto-import": "^20.2.0",
    "vite-plugin-glsl": "^1.5.1",
    "vite-plugin-qrcode": "^0.3.0",
    "vite-plugin-vue-devtools": "8.0.2",
    "vue-tsc": "^3.0.8"
  }
}
