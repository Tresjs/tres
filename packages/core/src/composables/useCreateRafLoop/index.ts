import { createEventHook, useRafFn } from '@vueuse/core'
import { Clock } from 'three'

export interface RafLoopContext { delta: number, elapsed: number }

/**
 * @param cycleFn the function that is called before the after event hook is triggered and after the before event hook is triggered.
 */
export const useCreateRafLoop = (cycleFn: () => void) => {
  const clock = new Clock()

  const eventHooks = {
    before: createEventHook<RafLoopContext>(),
    after: createEventHook<RafLoopContext>(),
  }

  const { pause, resume, isActive } = useRafFn(() => {
    const getContextWithClock = (): RafLoopContext => ({
      delta: clock.getDelta(),
      elapsed: clock.elapsedTime,
    })

    eventHooks.before.trigger(getContextWithClock())
    cycleFn()
    eventHooks.after.trigger(getContextWithClock())
  }, {
    immediate: false,
  })

  const start = () => {
    clock.start()
    resume()
  }

  const stop = () => {
    clock.stop()
    pause()
  }

  return {
    start,
    stop,
    isActive,
    onBeforeLoop: eventHooks.before.on,
    onLoop: eventHooks.after.on,
  }
}
