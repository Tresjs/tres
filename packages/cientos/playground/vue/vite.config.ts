import { templateCompilerOptions } from '@tresjs/core'
import vue from '@vitejs/plugin-vue'
import { resolve } from 'pathe'
import UnoCSS from 'unocss/vite'
import AutoImport from 'unplugin-auto-import/vite'
import Components from 'unplugin-vue-components/vite'
import { defineConfig } from 'vite'
import glsl from 'vite-plugin-glsl'
import { qrcode } from 'vite-plugin-qrcode'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [
    glsl(),
    vue(templateCompilerOptions),
    AutoImport({
      dts: true,
      eslintrc: {
        enabled: true, // <-- this
      },
      imports: ['vue'],
    }),
    Components({
      /* options */
    }),
    UnoCSS({
      theme: {
        colors: {
          'cientos-blue': '#82dbc5',
          'cientos-orange': '#fbb03b',
        },
      },
    }),
    qrcode(),
  ],
  resolve: {
    alias: {
      '@tresjs/cientos': resolve(__dirname, '../../src/'),
    },
  },
})
