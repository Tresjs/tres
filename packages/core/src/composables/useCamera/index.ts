import type { TresContext } from '../useTresContextProvider'

import type { ComputedRef, Ref } from 'vue'
import { computed, ref, watchEffect } from 'vue'
import { isCamera, isPerspectiveCamera } from '../../utils/is'
import type { TresCamera } from '../../types'

/**
 * Interface for the return value of the useCamera composable
 */
export interface UseCameraReturn {

  activeCamera: ComputedRef<TresCamera>
  /**
   * The list of cameras
   */
  cameras: Ref<TresCamera[]>
  /**
   * Register a camera
   * @param camera - The camera to register
   * @param active - Whether to set the camera as active
   */
  registerCamera: (camera: TresCamera, active?: boolean) => void
  /**
   * Deregister a camera
   * @param camera - The camera to deregister
   */
  deregisterCamera: (camera: TresCamera) => void
  /**
   * Set the active camera
   * @param cameraOrUuid - The camera or its UUID to set as active
   */
  setActiveCamera: (cameraOrUuid: string | TresCamera) => void
}

/**
 * Interface for the parameters of the useCamera composable
 */
interface UseCameraParams {
  sizes: TresContext['sizes']
}

/**
 * Composable for managing cameras in a Three.js scene
 * @param params - The parameters for the composable
 * @param params.sizes - The sizes object containing window dimensions
 * @returns The camera management functions and state
 */
export const useCameraManager = ({ sizes }: UseCameraParams): UseCameraReturn => {
  const cameras = ref<TresCamera[]>([])
  const activeCamera = computed<TresCamera>(() => cameras.value[0]) // the first camera is used to make sure there is always one camera active

  /**
   * Set the active camera
   * @param cameraOrUuid - The camera or its UUID to set as active
   */
  const setActiveCamera = (cameraOrUuid: string | TresCamera) => {
    const camera = isCamera(cameraOrUuid)
      ? cameraOrUuid
      : cameras.value.find((camera: TresCamera) => camera.uuid === cameraOrUuid)

    if (!camera) { return }

    const otherCameras = cameras.value.filter(({ uuid }) => uuid !== camera.uuid)
    cameras.value = [camera, ...otherCameras]
  }

  /**
   * Register a camera
   * @param camera - The camera to register
   * @param active - Whether to set the camera as active
   */
  const registerCamera = (camera: TresCamera, active = false): void => {
    if (cameras.value.some(({ uuid }) => uuid === camera.uuid)) { return }
    cameras.value.push(camera)

    if (active) {
      setActiveCamera(camera.uuid)
    }
  }

  /**
   * Deregister a camera
   * @param camera - The camera to deregister
   */
  const deregisterCamera = (camera: TresCamera): void => {
    cameras.value = cameras.value.filter(({ uuid }) => uuid !== camera.uuid)
  }

  /**
   * Update camera aspect ratios when the window size changes
   */
  watchEffect(() => {
    if (sizes.aspectRatio.value) {
      cameras.value.forEach((camera: TresCamera) => {
        if (isPerspectiveCamera(camera)) {
          camera.aspect = sizes.aspectRatio.value
          camera.updateProjectionMatrix()
        }
      })
    }
  })

  return {
    activeCamera,
    cameras,
    registerCamera,
    deregisterCamera,
    setActiveCamera,
  }
}
