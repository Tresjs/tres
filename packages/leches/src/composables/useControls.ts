import type { ToRefs } from 'vue'
import { isReactive, isRef, provide, reactive, ref, toRefs } from 'vue'
import type { Control } from '../types'

export const CONTROLS_CONTEXT_KEY = Symbol('CONTROLS_CONTEXT_KEY')
const DEFAULT_UUID = 'default'

// Internal state
const controlsStore: { [uuid: string]: { [key: string]: Control } } = reactive({})

export function useControlsProvider(uuid: string = DEFAULT_UUID) {
  provide(CONTROLS_CONTEXT_KEY, controlsStore)
  return controlsStore[uuid]
}
// Helper function to infer type
const inferType = (value: any): string => {
  const colorRegex = /^#(?:[A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$|^0x(?:[A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/

  if (typeof value === 'boolean') { return 'boolean' }
  if (typeof value === 'number') { return 'number' }
  if (typeof value === 'string' && colorRegex.test(value)) { return 'color' }
  if (typeof value === 'string') { return 'string' }
  if (value.isVector3
    || value.isVector2
    || value.isEuler
    || Array.isArray(value.value)
    || value.value.isVector3
    || value.value.isVector2
    || value.value.isEuler
    || Array.isArray(value.value.value)
  ) {
    return 'vector'
  }
  if (value.min !== undefined || value.max !== undefined || value.step !== undefined) { return 'range' }
  if (
    value.options
    && Array.isArray(value.options)) {
    return 'select'
  }

  // Add more types as needed
  return 'unknown'
}

const createControl = (key: string, value: any, type: string, folderName: string | null): Control => {
  const control: Control = {
    key: ref(key),
    label: ref(key),
    name: ref(key),
    type: ref(type),
    value: ref(value),
    visible: ref(true),
    icon: ref(),
    uniqueKey: ref(key),
    [key]: ref(value),
  }

  if (folderName) {
    control.folder = ref(folderName)
  }

  return control
}

export const dispose = (uuid: string = DEFAULT_UUID): void => {
  for (const key in controlsStore[uuid]) {
    delete controlsStore[uuid][key]
  }
}

export const useControls = (
  folderNameOrParams: string | { [key: string]: any },
  paramsOrOptions?: { [key: string]: any } | { uuid?: string },
  options?: { uuid?: string },
): Control | ToRefs<{ [key: string]: Control }> => {
  const result: { [key: string]: Control } = {}

  const folderName = typeof folderNameOrParams === 'string' ? folderNameOrParams : null
  const controlsParams = folderName ? paramsOrOptions as { [key: string]: any } : folderNameOrParams

  const actualOptions = folderName && folderName !== 'fpsgraph' ? options! : paramsOrOptions as { uuid?: string }
  const uuid = actualOptions?.uuid || DEFAULT_UUID

  if (!controlsStore[uuid]) {
    controlsStore[uuid] = reactive({})
  }

  if (folderNameOrParams === 'fpsgraph') {
    const control = createControl('fpsgraph', null, 'fpsgraph', null)
    controlsStore[uuid].fpsgraph = control
    result.fpsgraph = control
    return toRefs(reactive(result))
  }

  const controls = controlsStore[uuid]

  // Check if controlsParams is a reactive object
  const isParamsReactive = isReactive(controlsParams)
  const reactiveRefs = isParamsReactive ? toRefs(controlsParams as { [key: string]: any }) : {}

  for (const key in controlsParams as any) {
    let value = (controlsParams as any)[key]
    let uniqueKey = key

    // If controlsParams is reactive, use the reactive ref directly
    if (isParamsReactive && reactiveRefs[key]) {
      value = reactiveRefs[key]
    }

    // If the control is part of a folder, prefix the key with the folder name
    if (folderName) {
      uniqueKey = `${folderName}${key.charAt(0).toUpperCase() + key.slice(1)}`
    }

    // If the value is an object with control options
    if (typeof value === 'object' && !isRef(value) && !Array.isArray(value) && value.value !== undefined) {
      const controlOptions = value
      const reactiveValue = isRef(controlOptions.value) ? controlOptions.value : ref(controlOptions.value)
      const controlType = controlOptions.type || inferType(controlOptions)
      const control = createControl(key, reactiveValue, controlType, folderName)

      if (controlType === 'select') {
        control.options = ref(controlOptions.options.map((option: string | { text: string, value: any }) => {
          if (typeof option === 'object' && option.text && option.value) {
            return option
          }
          else {
            return {
              text: option,
              value: option,
            }
          }
        }))
      }

      if (controlType === 'range') {
        control.min = ref(controlOptions.min || 0)
        control.max = ref(controlOptions.max || 1)
        control.step = ref(controlOptions.step || 0.1)
      }

      control.label.value = controlOptions.label || key
      control.icon.value = controlOptions.icon
      control.visible.value = controlOptions.visible !== undefined ? controlOptions.visible : true
      control.uniqueKey.value = uniqueKey
      controls[uniqueKey] = control
      result[uniqueKey] = control
      continue
    }

    // If the value is a ref, use it directly
    if (isRef(value)) {
      const control = createControl(key, value, (value.value as any).type || inferType(value.value), folderName)
      controls[uniqueKey] = control
      result[uniqueKey] = control
      continue
    }

    // If the value is reactive, convert it to ref
    else if (typeof value === 'object' && !Array.isArray(value)) {
      const reactiveRefs = toRefs(value)
      if (reactiveRefs[key]) {
        value = reactiveRefs[key]
      }
    }

    // For non-ref values
    const control = createControl(key, value, value.type || inferType(value), folderName)

    // Update the internal state
    controls[uniqueKey] = control
    result[uniqueKey] = control
    control.uniqueKey.value = uniqueKey
  }

  return Object.keys(result).length > 1 ? toRefs(reactive(result)) : Object.values(result)[0]
}
