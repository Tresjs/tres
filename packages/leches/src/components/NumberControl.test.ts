import { mount } from '@vue/test-utils'
import { TresLeches, useControls } from '/@/'
import { beforeEach, describe, expect, it } from 'vitest'
import { defineComponent, nextTick } from 'vue'
import { dispose } from '/@/composables/useControls'

describe('number Control', async () => {
  let wrapper
  let component

  const mountComponent = (setup) => {
    component = defineComponent({
      template: `<TresLeches />`,
      components: { TresLeches },
      setup,
    })

    wrapper = mount(component, {
      components: { TresLeches },
    })
  }

  beforeEach(() => {
    dispose()
    mountComponent(() => {
      const { numberValue, visible } = useControls({ numberValue: 5 })
      return {
        numberValue,
        visible,
      }
    })
  })
  it('should render a number control', async () => {
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('input').attributes('type')).toBe('number')
  })
  it('should render a number control with a label', async () => {
    expect(wrapper.find('label').text()).toBe('numberValue')
  })
  it('should change the value of the control when the input changes', async () => {
    const input = wrapper.find('input[type="number"]')

    await input.setValue(20)
    await nextTick()

    expect(wrapper.vm.numberValue).toBe(20)
  })
  it('should hide the control when the visible property is toggled', async () => {
    wrapper.vm.visible = false
    await nextTick()
    expect(wrapper.find('input[type="number"]').exists()).toBe(false)
  })
  it('should show the control by default', async () => {
    const checkboxInput = wrapper.find('input[type="number"]')
    expect(checkboxInput.exists()).toBe(true)

    expect(wrapper.vm.visible).toBe(true)
  })
  it('should not show the control if the visibility is initially false', async () => {
    dispose()
    mountComponent(() => {
      const { numberValue, visible } = useControls({ numberValue: { value: 5, visible: false } })
      return {
        numberValue,
        visible,
      }
    })
    const checkboxInput = wrapper.find('input[type="number"]')
    expect(checkboxInput.exists()).toBe(false)

    expect(wrapper.vm.visible).toBe(false)
  })
  it('should render an icon instead of a label', () => {
    dispose()
    mountComponent(() => {
      const { visible } = useControls({
        test: {
          value: 2,
          icon: 'i-carbon-checkmark',
        },
      })
      return { visible }
    })
    const icon = wrapper.find('i.i-carbon-checkmark')
    expect(icon.exists()).toBe(true)
  })
})
