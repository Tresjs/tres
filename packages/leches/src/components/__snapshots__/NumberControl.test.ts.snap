// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Number Control > should render a number control 1`] = `
"<div style=\\"left: 704px; top: 10px;\\" class=\\"tl-absolute tl-select-none tl-z-24 tl-w-280px tl-font-sans tl-text-xs\\">
  <div tabindex=\\"0\\" class=\\"tl-bg-white tl-shadow-xl tl-rounded tl-border-4 tl-border-solid tl-border-black\\">
    <header class=\\"tl-relative tl-cursor-grabbing tl-p-4 tl-flex tl-justify-between tl-text-gray-200 tl-relative\\"><i class=\\"tl-h-4 tl-w-4 tl-p-1.5 tl-flex tl-items-center tl-line-height-0 tl-rounded-full tl-bg-gray-100 tl-text-xs\\">🍰</i>
      <div><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i></div>
      <div></div>
    </header>
    <!--v-if-->
    <div>
      <div class=\\"tl-flex tl-px-4 tl-justify-between tl-gap-4 tl-items-center tl-mb-2\\"><label for=\\"numberValue\\" title=\\"numberValue\\" class=\\"tl-text-gray-500 tl-w-1/3 tl-truncate\\">numberValue</label><input id=\\"numberValue\\" class=\\"tl-p-2 tl-w-1/3 tl-rounded tl-text-right tl-text-xs tl-text-gray-400 tl-bg-gray-100 tl-focus:border-gray-200 tl-outline-none tl-border-none tl-font-sans\\" type=\\"number\\"></div>
    </div>
  </div>
</div>"
`;

exports[`number Control > should render a number control 1`] = `
"<div style="left: 704px; top: 10px;" class="tl-absolute tl-select-none tl-z-24 tl-w-280px tl-font-sans tl-text-xs">
  <div tabindex="0" class="tl-bg-white tl-shadow-xl tl-rounded tl-border-4 tl-border-solid tl-border-black">
    <header class="tl-relative tl-cursor-grabbing tl-p-4 tl-flex tl-justify-between tl-text-gray-200 tl-relative"><i class="tl-h-4 tl-w-4 tl-p-1.5 tl-flex tl-items-center tl-line-height-0 tl-rounded-full tl-bg-gray-100 tl-text-xs">🍰</i>
      <div><i class="i-ic-baseline-drag-indicator"></i><i class="i-ic-baseline-drag-indicator"></i><i class="i-ic-baseline-drag-indicator"></i></div>
      <div></div>
    </header>
    <!--v-if-->
    <div>
      <div class="tl-flex tl-px-4 tl-justify-between tl-gap-4 tl-items-center tl-mb-2"><label for="numberValue" title="numberValue" class="tl-text-gray-500 tl-w-1/3 tl-truncate">numberValue</label><input id="numberValue" class="tl-p-2 tl-w-1/3 tl-rounded tl-text-right tl-text-xs tl-text-gray-400 tl-bg-gray-100 tl-focus:border-gray-200 tl-outline-none tl-border-none tl-font-sans" type="number" value="5"></div>
    </div>
  </div>
</div>"
`;
